src/d3a/models/myco_matcher/external_matcher.py

```diff
@@ -145,38 +140,36 @@ def _get_bids_offers(market: Market, filters: Dict):
                 offer.serializable_dict() for offer in offers.values())
         return bids_list, offers_list
 
-    def _get_validated_bid_offer_match_list(
+    def _get_validated_recommendations(
             self, recommendations: List[Dict]) -> List[BidOfferMatch.serializable_dict]:
         """Return a validated list of BidOfferMatch instances."""
-        validated_records = []
-        for record in recommendations:
-            market = self.markets_mapping.get(record.get("market_id"), None)
+        validated_recommendations = []
+        for recommendation in recommendations:
+            market = self.markets_mapping.get(recommendation.get("market_id"))
             if market is None or market.readonly:
                 # The market is already finished or doesn't exist
                 raise MycoValidationException
 
             # Get the original bid and offer from the market
-            market_bid = market.bids.get(record.get("bid").get("id"), None)
-            market_offer = market.offers.get(record.get("offer").get("id"), None)
+            market_bids = [
+                market.bids.get(bid.get("id"), None) for bid in recommendation.get("bids")]
+            market_offers = [
+                market.offers.get(offer.get("id"), None) for offer in recommendation.get("offers")]
 
-            if not (market_bid and market_offer):
-                # Offer or Bid either don't belong to market or were already matched
+            if not (all(market_bids) and all(market_offers)):
```

Comment:
Why is `all()` needed here? An empty list evaluates to `False`, so this should also work:
```suggestion
            if not (market_bids and market_offers):
```

Reply:
I'm checking here a different case, not if we have an empty list [] , I'm checking the case when we have [offer1, offer2, None, offer3]
If some offer doesn't exist we will have  None , which is the case I'm testing here

Reply:
And how does a `None` end up in `recommendation["offers"]` (or bids)?

Reply:
we don't, we get them in the market bids/offers in this line `market_offers = [
                market.offers.get(offer.get("id")) for offer in recommendation.get("offers")]`
 market.offers.get(offer.get("id"))  could return None if the offer id doesn't exist in the market, and then I'm checking `if not (all(market_bids) and all(market_offers)):`

Reply:
Ah yes, I see it now. Thanks

Reply:
Another question: why do you use `offer.get("id")` rather than `offer["id"]`? In which case can we have an offer without ID?

Reply:
The data is received from a client side, there should be a validation class-function imo to make sure the format is supported, anw I will not add this in my current PR

Reply:
I suggest creating a Jira task for it then, since it looks quite an important requirement 👍

Reply:
https://gridsingularity.atlassian.net/browse/D3ASIM-3542